### ISSUE TITLE
flux job eventlog/wait-event: pick main vs exec eventlog

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#2105]
need a mechanism to pick an eventlog.  `--primary | --exec`?  Or if there are future possible eventlogs, perhaps ```--which <primary | exec>``` (I know `--which` is terrible, can't think of a good option name right now).

**Comment by grondo** [comment]
Good question. I can't really think of a great interface for `eventlog`, that is a tricky one.

For `wait-event` what if we support an `exec` prefix for event names, e.g.

```
flux job wait-event exec.running
```

But maybe that is too ad hoc?

Similarly, what if `flux job eventlog` by default dumped both eventlogs, sorted by timestamp?

Sorry I'm just kind of thinking out loud. I'm also wondering if having two separate eventlogs was a misstep that is overcomplicating our required interfaces.

**Comment by chu11** [comment]
> For wait-event what if we support an exec prefix for event names

So the prefix would indicate which eventlog?  So like `primary.submit` would be for the primary eventlog?  Of course no prefix defaults to the primary one.

It works, but does seem a little too ... esoteric?  An average user I think would just expect an option?

> Similarly, what if flux job eventlog by default dumped both eventlogs, sorted by timestamp?

I didn't even think about that.  But how would `flux job wait-event` work.  check both eventlogs?

**Comment by grondo** [comment]
> It works, but does seem a little too ... esoteric? An average user I think would just expect an option?

I was kind of hoping to hide the fact that there are two disjoint eventlogs from users, however, perhaps you are right trying to hide that fact too much might just get us into trouble. I really don't like how the interface is turning out due to the split events though...

> I didn't even think about that. But how would flux job wait-event work. check both eventlogs?

Both `eventlog` and `wait-event` would prefix `guest.exec.eventlog` events with `exec.` -- i.e. other supported eventlogs are in a different namespace... Like I said though, more thinking out loud at this point.

**Comment by garlick** [comment]
I tend to side with @chu11 on not hiding the fact that there are two eventlogs.  Interface-wise, maybe just have the primary be the default path, but allow it to be overridden with something like:
```
flux job eventlog [--path exec.eventlog] ID
flux job wait-event [--path exec.eventlog] ID event
```

The larger question of whether this design (two/multiple eventlogs) is the right one seems like should have some discussion.  We might want to revisit the underlying assumptions and motivations of this design and see if they still hold now that other components of the system have become more concrete.

**Comment by grondo** [comment]
Do you mean
```
[--path guest.exec.eventlog]
```

I don't disagree that we perhaps shouldn't hide the two separate eventlogs from users. However, we already have to add some "magic" to avoid races when watching the `guest.exec.eventlog` and requiring a user to know the "path" to the exec eventlog to match events there is not a great design IMO.

What if we allow a small set of `exec` system events to be put into the primary eventlog, choosing those that are likely to require synchronization. Perhaps starting with
```
exec.running  # all tasks running
exec.stopped  # all tasks stopped waiting for debugger
```

The job shell could optionally create its own eventlog if that is required for its own purposes, and then since only shell related events are in that log, it isn't quite as onerous to require user of the shell to know an specify the path to the eventlog. (and perhaps it would only be used post-mortem or for informational purposes).

Just keeping the ideas flowing. I'm fine with going with group consensus for now if that is the way forward.

**Comment by grondo** [comment]
Imagine a use case of a tool that needs to synchronize on the stopped state of a job. How will that work in the current case? Should we devise a few common use cases like this, and create an interface that works easily for those use cases, then figure out the best implementation for that interface?  It feels a bit like we're doing it the other way around right now, but I'm probably missing something.

**Comment by garlick** [comment]
Yeah, why don't we try it that way.  Since the events you mentioned above originate with the job shell running as the user, how would we route them to the log?

One thought is to implement a (simple) method in the job manager that allows a job owner to add events (an authorized set of event names?  names with an approved prefix?)  to its eventlog.  The job manager could trivially check requestor against job owner and validate that job is active, then drop these events in a timed KVS batch, responding after the commit is completed.

Then for now we could just support fetching the guest eventlog, not synchronizing on it?

**Comment by grondo** [comment]
I had figured the exec system would send these `exec` events back to the job-manager. I had in mind that the exec system could either monitor the `guest.exec.eventlog` and shuttle events of interest back to the job-manager, or we could define a protocol over the job shell's stdio (or perhaps another defined fd passed to the job shells) where the job shells could send events back to the exec module.
(BTW, your solution above sounds fine too, and seems to remove the middle man.)

Admittedly, I hadn't thought this through in detail, and now I see that we're just trading one complexity for another, which is probably the reason we split the eventlogs in twain in the first place. I apologize for wasting everyone's time.

**Comment by garlick** [comment]
We could do it that way too if that makes more sense (it was what I was originally thinking).

I think the idea of using only the main eventlog for synchronization could have a simplifying effect so maybe some more discussion is warranted?

**Comment by chu11** [comment]
Just documenting tricky / racy things related to reading an eventlog from the guest namespace.

- if the guest namespace eventlog is already copied to the main namespace ("release" state and "final=true" in main eventlog), this is easy, same logic as current code except a different eventlog.  This is the easy case.

- if the guest eventlog is still active ("start" state in main eventlog, but not released), need to watch from the guest namespace instead of the main one.

  - racy component - attempting to watch from the guest namespace, guest namespace may have been released by time request for watching it occurred.  Need to handle and go back to "copied into main namespace" case.

  - tricky component - if guest namespace is released while watching it (presumably have seen some events fly by), and the user has not canceled, should convert to watching the same eventlog after it has been copied to main namespace.  And should do offset magic to ensure duplicate events aren't sent.

- if guest namespace has not yet been created ("start" state has not yet happened), need to wait for the namespace to be created before watching it.  Then do the above cases.

As discussed in a thread on slack, the initial plan of adding guest eventlog support into job-info may not be a good idea given all the trickiness above.  May be best to create a new module callback that works on top of the current job info lookup & watch to handle guest eventlogs.  We can hide which callback is called via `flux_job_event_watch()`.

**Comment by chu11** [comment]
closed via #2307
